File: Lab1/mm.py
# Another migration code written and revised by Daniel Yeung, Python matrix multiplication version 1.01, adopted from Prof Cao adapting from https://csapp.cs.cmu.edu/3e/ics3/code/mem/matmult/mm.c 
import time
import sys

# As a good habit learnt from COMP1002, we write docstrings in functions/procedures

MAXN = 1024

# Initializing global matrix constants with all elements set to 0
ga = [[0.0 for _ in range(MAXN)] for _ in range(MAXN)]
gb = [[0.0 for _ in range(MAXN)] for _ in range(MAXN)]
gc = [[0.0 for _ in range(MAXN)] for _ in range(MAXN)]

# Unlike C and Java, there is no need for declaring a function pointer (Dynamic typing!)

def check_result(c, n):
    """multiplication check for correct results."""
    for i in range(n):
        for j in range(n):
            if c[i][j] != float(n):
                print("Error: bad number ({:.1f}) in result matrix ({},{})".format(c[i][j], i, j))
                sys.exit(0)

def run(func, n):
    """simply runs the program, more importantly we care about time*"""
    start_time = time.time()
    func(ga, gb, gc, n)
    end_time = time.time()
    check_result(gc, n)
    return end_time - start_time

def init(a, b, n):
    """set every elements in arrays as 1."""
    for i in range(n):
        for j in range(n):
            a[i][j] = 1.0
            b[i][j] = 1.0

def ijk(A, B, C, n):
    """ijk way"""
    for i in range(n):
        for j in range(n):
            sum = 0.0
            for k in range(n):
                sum += A[i][k] * B[k][j]
            C[i][j] += sum

def jik(A, B, C, n):
    """jik way"""
    for j in range(n):
        for i in range(n):
            sum = 0.0
            for k in range(n):
                sum += A[i][k] * B[k][j]
            C[i][j] += sum

def ikj(A, B, C, n):
    """ikj way"""
    for i in range(n):
        for k in range(n):
            r = A[i][k]
            for j in range(n):
                C[i][j] += r * B[k][j]

def kij(A, B, C, n):
    """kij way"""
    for k in range(n):
        for i in range(n):
            r = A[i][k]
            for j in range(n):
                C[i][j] += r * B[k][j]

def kji(A, B, C, n):
    """kji way"""
    for k in range(n):
        for j in range(n):
            r = B[k][j]
            for i in range(n):
                C[i][j] += A[i][k] * r

def jki(A, B, C, n):
    """jki way"""
    for j in range(n):
        for k in range(n):
            r = B[k][j]
            for i in range(n):
                C[i][j] += A[i][k] * r

def print_array(A, n):
    """Nope."""
    print("array")

def usage():
    """reminds that you didn't parse anything"""
    print("Usage: python mm.py <ijk|jik|jki|kji|kij|ikj>")

def main():
    if len(sys.argv) < 2:
        usage()
        return

    option = 0
    for i in range(3):
        if i < len(sys.argv[1]):
            option = (option << 1) + ord(sys.argv[1][i]) - ord('i')

    if option < 4 or option > 10:
        usage()
        return

    n = MAXN
    init(ga, gb, n)

    total_t = 0
    if option == 4:
        total_t = run(ijk, n)
    elif option == 5:
        total_t = run(ikj, n)
    elif option == 6:
        total_t = run(jik, n)
    elif option == 8:
        total_t = run(jki, n)
    elif option == 9:
        total_t = run(kij, n)
    elif option == 10:
        total_t = run(kji, n)
    
    print("It takes {:.3f} seconds to multiply two {}*{} double matrices.".format(total_t, n, n))

if __name__ == "__main__":
    main()

File: Lab1/mm.java
// Java matrix multiplication version 1.01 , revised by Daniel Yeung, adopted from Prof Cao adapting from https://csapp.cs.cmu.edu/3e/ics3/code/mem/matmult/mm.c */
public class mm {
    // constants initialization
    private static final int MAXN = 1024;
    private static double[][] ga = new double[MAXN][MAXN];
    private static double[][] gb = new double[MAXN][MAXN];
    private static double[][] gc = new double[MAXN][MAXN];

    // equivalent to test_funct from C
    @FunctionalInterface
    interface TestFunction {
        void test(double[][] A, double[][] B, double[][] C, int n);
    }

    // stubbornly included function, finds out if there is wrong multiplication the computer has made
    private static void checkResult(double[][] c, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (c[i][j] != (double) n) {
                    System.out.printf("Error: bad number (%f) in result matrix (%d,%d)%n", c[i][j], i, j);
                    System.exit(0);
                }
            }
        }
    }

    // the soul of this program, precisely capturing time required for the matrix multiplications
    private static double run(TestFunction f, int n) {
        long startTime = System.nanoTime();
        f.test(ga, gb, gc, n);
        long endTime = System.nanoTime();
        checkResult(gc, n);
        return (endTime - startTime) / 1_000_000_000.0;  // 1 ns  = 10^-9 s
    }

    // no surprises! All elements for the arrays are 1
    private static void init(double[][] a, double[][] b, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                a[i][j] = 1.0;
                b[i][j] = 1.0;
            }
        }
    }

    // Daniel doesn't like debugging, hence, the printArray is omitted
    // Please implement it yourself if you intend to have one.
    private static void printArray(double[][] a, int n){
        // Refer to Recording 1 : 16:07 - 16:40
        System.out.printf("We do not debug algorithms - Professor Cao 16/1/2025 08:46:17");
    }

    // pov : you tried to run this program without giving it params
    private static void usage() {
        System.out.println("Usage: java mm <ijk|jik|jki|kji|kij|ikj>");
    }

    // the six similar pieces of code with permutations of ijk
    private static void ijk(double[][] A, double[][] B, double[][] C, int n) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                double sum = 0.0;
                for (int k = 0; k < n; k++) {
                    sum += A[i][k] * B[k][j];
                }
                C[i][j] += sum;
            }
        }
    }

    private static void jik(double[][] A, double[][] B, double[][] C, int n) {
        for (int j = 0; j < n; j++) {
            for (int i = 0; i < n; i++) {
                double sum = 0.0;
                for (int k = 0; k < n; k++) {
                    sum += A[i][k] * B[k][j];
                }
                C[i][j] += sum;
            }
        }
    }

    private static void ikj(double[][] A, double[][] B, double[][] C, int n) {
        for (int i = 0; i < n; i++) {
            for (int k = 0; k < n; k++) {
                double r = A[i][k];
                for (int j = 0; j < n; j++) {
                    C[i][j] += r * B[k][j];
                }
            }
        }
    }

    private static void kij(double[][] A, double[][] B, double[][] C, int n) {
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                double r = A[i][k];
                for (int j = 0; j < n; j++) {
                    C[i][j] += r * B[k][j];
                }
            }
        }
    }

    private static void kji(double[][] A, double[][] B, double[][] C, int n) {
        for (int k = 0; k < n; k++) {
            for (int j = 0; j < n; j++) {
                double r = B[k][j];
                for (int i = 0; i < n; i++) {
                    C[i][j] += A[i][k] * r;
                }
            }
        }
    }

    private static void jki(double[][] A, double[][] B, double[][] C, int n) {
        for (int j = 0; j < n; j++) {
            for (int k = 0; k < n; k++) {
                double r = B[k][j];
                for (int i = 0; i < n; i++) {
                    C[i][j] += A[i][k] * r;
                }
            }
        }
    }


    // main part, has some slight variations from og code, please revise if buggy

    public static void main(String[] args) {
        if (args.length < 1) {
            usage();
            return;
        }

        char option = 0;
        for (int i = 0; i < 3 && i < args[0].length(); i++) {
            option = (char) ((option << 1) + args[0].charAt(i) - 'i');
        }

        if (option < 4 || option > 10) {
            usage();
            return;
        }

        int n = MAXN;
        init(ga, gb, n);

        double total_t;
        switch (option) {
            case 4:
                total_t = run(mm::ijk, n);
                break;
            case 5:
                total_t = run(mm::ikj, n);
                break;
            case 6:
                total_t = run(mm::jik, n);
                break;
            case 8:
                total_t = run(mm::jki, n);
                break;
            case 9:
                total_t = run(mm::kij, n);
                break;
            case 10:
                total_t = run(mm::kji, n);
                break;
            default:
                usage();
                return;
        }
        System.out.printf("It takes %.3f seconds to multiply two %d*%d double matrices.%n", total_t, n, n);
    }
}